docs/root/intro/life_of_a_request.rst

```diff
@@ -0,0 +1,631 @@
+
+.. image:: /_static/lor-network-write.svg
+   :width: 80%
+   :align: center
+
+.. _life_of_a_request_http2_decoding:
+
+5. HTTP/2 codec decoding
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+The HTTP/2 codec in Envoy is based on `nghttp2 <https://nghttp2.org/>`_. It is invoked by HCM with
+plaintext bytes from the TCP connection (after network filter chain transformation). The codec
+decodes the byte stream as a series of HTTP/2 frames and demultiplexes the connection into a number
+of independent HTTP streams. Stream multiplexing is a key feature in HTTP/2, providing significant
+performance advantages over HTTP/1. Each HTTP stream handles a single request and response.
+
+The codec is also responsible for handling HTTP/2 setting frames and both stream and connection
+level flow control.
+
+The codecs are responsible for abstracting the specifics of the HTTP connection, presenting a
+standard view to the HTTP connection manager and filter chain of a connection split into streams,
+each with request/response headers/body/trailers. This is true regardless of whether the protocol is
+HTTP/1, HTTP/2 or HTTP/3.
+
+6. HTTP filter chain processing
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+For each HTTP stream, HCM instanties an :ref:`HTTP filter <arch_overview_http_filters>` chain,
+following the pattern established above for listener and network filter chains.
+
+.. image:: /_static/lor-http-filters.svg
+   :width: 80%
+   :align: center
+
+There are three kinds of HTTP filter interfaces:
+
+* :repo:`StreamDecoderFilter <include/envoy/http/filter.h>` with callbacks for request processing.
+* :repo:`StreamEncoderFilter <include/envoy/http/filter.h>` with callbacks for response processing.
+* :repo:`StreamFilter <include/envoy/http/filter.h>` implementing both `StreamDecoderFilter` and
+  `StreamEncoderFilter`.
+
+Looking at the decoder filter interface:
+
+.. code-block:: cpp
+
+  virtual FilterHeadersStatus decodeHeaders(RequestHeaderMap& headers, bool end_stream) PURE;
+  virtual FilterDataStatus decodeData(Buffer::Instance& data, bool end_stream) PURE;
+  virtual FilterTrailersStatus decodeTrailers(RequestTrailerMap& trailers) PURE;
+
+Rather than operating on connection buffers and events, HTTP filters follow the lifecycle of an HTTP
+request, e.g. ``decodeHeaders()`` takes HTTP headers as an argument rather than a byte buffer. The
+returned ``FilterStatus`` provides, as with network and listener filters, the ability to manage filter
+chain control flow.
+
+When the HTTP/2 codec makes available the HTTP requests headers, these are first passed to
+``decodeHeaders()`` in CustomFilter. If this ``FilterHeadersStatus`` is ``Continue``, HCM then
+passes the headers (possibly mutated by CustomFilter) to the router filter.
+
+Decoder and encoder-decoder filters are executed on the request path. Encoder and encoder-decoder
+filters are executed on the response path, in reverse direction. In both paths, the same filter
```

Comment:
I think the second sentence about the same "ordering" is confusing. I see what you are saying: that the ordering is not changed, just reverse, but I think that using the diagrams below and deleting this sentence is sufficient.